src/arkode/arkode_sprkstep.c

```diff
@@ -0,0 +1,786 @@
+  of the subsequent time step.
+
+  If it is called in ARK_FULLRHS_END mode and the method coefficients
+  support it, we may just copy the stage vectors to fill f instead
+  of calling f().
+
+  ARK_FULLRHS_OTHER mode is only called for dense output in-between
+  steps, so we strive to store the intermediate parts so that they
+  do not interfere with the other two modes.
+  ---------------------------------------------------------------*/
+int sprkStep_FullRHS(void* arkode_mem, realtype t, N_Vector y, N_Vector f,
+                     int mode)
+{
+  int retval                 = 0;
+  ARKodeMem ark_mem          = NULL;
+  ARKodeSPRKStepMem step_mem = NULL;
+
+  /* access ARKodeSPRKStepMem structure */
+  retval = sprkStep_AccessStepMem(arkode_mem, "SPRKStep_FullRHS", &ark_mem,
+                                  &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* perform RHS functions contingent on 'mode' argument */
+  switch (mode)
+  {
+  case ARK_FULLRHS_START:
+  case ARK_FULLRHS_END:
+  case ARK_FULLRHS_OTHER:
+
+    retval = sprkStep_f1(step_mem, t, y, f, ark_mem->user_data);
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, "ARKODE::SPRKStep",
+                      "SPRKStep_FullRHS", MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
+
+    retval = sprkStep_f2(step_mem, t, y, f, ark_mem->user_data);
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, "ARKODE::SPRKStep",
+                      "SPRKStep_FullRHS", MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
+
+    break;
+
+  default:
+    /* return with RHS failure if unknown mode is passed */
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, "ARKODE::SPRKStep",
+                    "SPRKStep_FullRHS", "Unknown full RHS mode");
+    return (ARK_RHSFUNC_FAIL);
+  }
+
+  return (ARK_SUCCESS);
+}
+
+/* Standard formulation of SPRK.
+   This requires only 2 vectors in principle, but we use three
+   since we persist the stage data. Only the stage data vector
+   belongs to SPRKStep, the other two are reused from the ARKODE core. */
+int sprkStep_TakeStep(void* arkode_mem, realtype* dsmPtr, int* nflagPtr)
+{
+  ARKodeMem ark_mem          = NULL;
+  ARKodeSPRKStepMem step_mem = NULL;
+  N_Vector prev_stage        = NULL;
+  N_Vector curr_stage        = NULL;
+  sunrealtype ci = 0.0;
+  sunrealtype chati = 0.0;
+  int is                     = 0;
+  int retval                 = 0;
+
+  /* access ARKodeSPRKStepMem structure */
+  retval = sprkStep_AccessStepMem(arkode_mem, "sprkStep_TakeStep", &ark_mem,
+                                  &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  prev_stage    = ark_mem->yn;
+  curr_stage    = ark_mem->ycur;
+  for (is = 0; is < step_mem->method->stages; is++)
+  {
+    /* load/compute coefficients */
+    sunrealtype ai    = step_mem->method->a[is];
+    sunrealtype ahati = step_mem->method->ahat[is];
```

Comment:
This is not a mixed declaration since its at the top of the block (it passed when I compiled with C90, all warnings, and warnings as errors).

Reply:
Yes, you're right it's at the top of the block so I think it's ok. It just stood out to me because I think this is the only place where everything is not declared at the top of the function.